### PR TITLE
feat(#910): extract agent-generic turn-driver into amplihack-turn crate (PR-2)

### PR DESCRIPTION
**Pull request by rysweet**
## Summary (PR-2 of a multi-PR #910 refactor)

Adds the new agent-generic `amplihack-turn` crate that houses the reusable turn-driver abstraction for issue #910. **Behavior-preserving** — no signal or auto-mode behavior changes; the PR-1 characterization tests pass unchanged.

## What this PR does
- New leaf crate `crates/amplihack-turn`:
  - `AgentSession` trait (`run_turn`, resumable; runs to natural completion, no wall-clock cap).
  - `Channel` trait (`publish_output` = REPLAY default no-op; `next_prompt` = LISTEN) + `NextPrompt::{Ready,Idle,Closed}`.
  - Generic `run_session_loop(session, channel)` — the one place the drive-a-turn loop lives.
  - Relocated `SerialTurnDriver` / `CopilotTurnRunner` / `build_turn_argv` (+ its `ToolAllowlist` dependency) out of `amplihack-signal` (they were already agent-generic).
- `amplihack-signal` now depends on `amplihack-turn` (optional, under the `signal` feature) and re-exports the moved items via `pub use amplihack_turn::{...}`, so its API and tests are unchanged.
- Hermetic mock-session/mock-channel driver-loop contract tests (no network, no real copilot).
- API reference doc.

## Crusty condition #4 (net-gating) — satisfied
`amplihack-turn` uses `tokio` with `default-features = false` and only `process/rt/macros/time/sync/io-util` — **never `net`**. Non-signal builds stay free of the tokio net stack; the net stack remains gated in signal's `transport.rs`.

## Scope
Does NOT rewire `run_chat_async` or auto-mode to use the loop — that is PR-3 (SignalChannel + delete dead `SignalSession`/`Inbox`) and PR-4 (AutoModeChannel). Refs #910 (does not close it).

### CODE REVIEW

File: crates/amplihack-turn/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            NextPrompt::Ready(prompt) => {
+                // Run to natural completion; a TurnError converts into a
+                // ChannelError via `?` (From) and propagates unchanged.
+                let out = session.run_turn(&prompt).await?;
+                // REPLAY (default no-op). A publish failure fails fast.
+                channel.publish_output(&out).await?;
+            }
+            // Nothing to run yet: yield to the runtime and re-poll. No timeout,
+            // no busy-spin. A correct channel returns Idle only transiently.
+            NextPrompt::Idle => tokio::task::yield_now().await,
```
Comment (rysweet) [inline crates/amplihack-turn/src/lib.rs]: **P2 — design concern (latent): Idle handling can busy-spin.**

The doc comment states the wait "yields cooperatively to the runtime rather than busy-spinning the CPU," but `tokio::task::yield_now()` does **not** block — it re-queues the task and resumes on the next scheduler tick. If a `Channel` returns `NextPrompt::Idle` repeatedly (which the trait explicitly permits: *"Nothing to run yet"*), this loop becomes a hot poll spinning `next_prompt()` at ~100% CPU. The anti-spin guarantee is therefore delegated entirely to every `Channel` impl ("returns Idle only transiently").

This is latent right now (no channel is wired to the loop in this PR), so it is **not blocking** for a move+abstraction PR. But before a real channel adopts `run_session_loop`, either:
1. Tighten the `next_prompt` contract so it MUST block internally on liveness/inbound and only returns `Idle` for genuine spurious wakeups, and document that; or
2. Add a bounded backoff (e.g. a short `tokio::time::sleep`) in the `Idle` arm.

As written, a well-meaning but naive channel that returns `Idle` in a tight loop will peg a core.

**Review by rysweet — COMMENTED** [review]
## Step 17b — Comprehensive Code Review

**Verdict: APPROVE with comments.** This is a clean, behaviour-preserving extraction. All validation gates pass locally; no blocking (P0) issues found. Findings below are one latent design concern (P2) and minor advisories (P3).

### What I verified

| Check | Command | Result |
|---|---|---|
| New crate builds + contract tests | `cargo test -p amplihack-turn` | ✅ 14/14 |
| PR-1 characterization tests unchanged | `cargo test -p amplihack-signal --features signal` | ✅ all suites green |
| Default (non-signal) build | `cargo build -p amplihack-signal` | ✅ compiles (shims gated behind `signal`) |
| Formatting | `cargo fmt --all --check` | ✅ clean |
| Lint | `cargo clippy -p amplihack-turn --all-targets` | ✅ no warnings |
| Feature footprint (crusty #4) | `cargo tree -p amplihack-turn -e features -i tokio` | ✅ `process, rt, macros, time, sync, io-util` — **no `net`** |

### Move fidelity (behaviour-identical) — confirmed
- `turn.rs`: byte-for-byte identical to `main`'s `chat/turn.rs` except the single required import change `use super::allowlist::ToolAllowlist;` → `use crate::allowlist::ToolAllowlist;`.
- `allowlist.rs`: **zero** diff against `main`'s `chat/allowlist.rs`.
- Shims (`chat/turn.rs`, `chat/allowlist.rs`) re-export the relocated items at their original paths, preserving every public signature. Wiring is correct: `amplihack-turn` is an **optional** dep enabled only by the `signal` feature, and the `chat` module is itself `#[cfg(feature = "signal")]`, so default builds never reference it.

### Review checklist
- [x] Code quality and standards — idiomatic, well-documented, `#[must_use]` where appropriate.
- [x] Test coverage adequate — 14 hermetic tests cover run→publish ordering, prompt ordering, session-id reuse, `Idle` wait, `Closed` break, default no-op publish, and **all three** error-propagation paths (turn/publish/recv), plus object-safety via `dyn Channel` and `From` conversions.
- [x] No TODOs, stubs, or swallowed exceptions — every failure surfaces via `thiserror` enums and propagates through `?`; no silent fallback.
- [x] No unimplemented functions.
- [x] Logic correctness — driver loop matches its documented contract and the tests.
- [x] Edge case handling — errors, idle, closed, and no-op publish all covered.

### Findings

**P2 — latent design concern: `Idle` handling can busy-spin.** See inline comment on `lib.rs:215`. `tokio::task::yield_now()` does not block; a `Channel` that returns `NextPrompt::Idle` repeatedly (which the trait permits) makes the loop a ~100% CPU hot-poll. Not blocking for this move/abstraction PR (no channel is wired yet), but the `next_prompt` contract should require internal blocking on liveness, or the `Idle` arm should add a bounded backoff, before a real channel adopts `run_session_loop`. The doc comment's "not busy-spin" claim is currently aspirational, not enforced.

**P3 — workspace version bump (`0.12.0` → `0.13.0`) in `Cargo.toml` + `package.json`.** Reasonable for adding a public crate, but version bumps inside a feature PR can conflict with other concurrent PRs and are usually done at release time. Flagging for awareness; no change required.

**P3 — pre-existing (out of scope): `turn.rs:234`** embeds child `stdout`+`stderr` into the non-zero-exit error string. Behaviour-identical move; correctly not touched here. Consider addressing in a follow-up (avoid leaking full child output into an error string), not in this move-only PR.

**ℹ️ Info — branch is behind `main`.** The merge base is the PR-1 commit; `merge-flow.md` (issue #1050) was added to `main` *after* the branch point. GitHub's two-dot diff therefore shows `merge-flow.md` as *deleted* and an extra `docs/index.md` hunk, but the PR's true delta (`merge-base..HEAD`) does **not** delete it — a 3-way merge preserves the file and both `index.md` additions land in different regions (low conflict risk). Recommend rebasing/updating on `main` so the displayed diff accurately reflects the real change. Also consider renaming the auto-generated title *"Update Cargo.toml with 14 changed files (#910)"* to describe the extraction.

**Bottom line:** Solid, well-tested, behaviour-preserving refactor that cleanly resolves the allowlist coupling and keeps the always-compiled crate net-free. Approve; address the P2 before wiring a live channel to the loop.

**Comment by rysweet** [comment]
## Step 17c — Security Review

**Verdict: APPROVE (no security-blocking issues).** This is a behaviour-identical extraction of the turn-driver/allowlist into a new `amplihack-turn` crate; the security posture is **preserved unchanged**, with no new vulnerabilities introduced.

### Scope verified (move fidelity)
- `turn.rs`: only diff vs `main` is the import path `super::allowlist` → `crate::allowlist`. Logic byte-for-byte identical.
- `allowlist.rs`: **byte-identical** to `main`.
- `amplihack-signal/src/chat/{turn,allowlist}.rs`: reduced to re-export shims (original public paths preserved).
- No new third-party dependencies: the only crate added to `Cargo.lock` is the new workspace member `amplihack-turn`; its deps (`tokio` subset, `async-trait`, `thiserror`) already exist in the workspace.

### Security checklist
| Area | Result |
|------|--------|
| **Injection (command/shell)** | ✅ Safe. `build_turn_argv` passes the attacker-influenced prompt as **exactly one argv element** via `Command::args()` — no shell, no string concatenation. Metacharacters cannot inject a command. |
| **AuthZ / least-privilege** | ✅ Preserved. Default is read-only (`view`/`grep`/`glob`); blanket access requires explicit `--dangerous-all-tools` and maps to Copilot's `--allow-all-tools` (tools-only), **never** the wider `--allow-all`. Byte-identical to `main`. |
| **Process lifecycle / TOCTOU** | ✅ Strong. Pre-emption fires a child-bound one-shot and kills the **owned** `tokio::process::Child` via `start_kill` — no raw PID is ever exposed, eliminating the PID-reuse race. `stdin` is `null`; `kill_on_drop(true)` prevents orphaned children on cancellation/unwind. |
| **Attack surface (net)** | ✅ Not expanded. `amplihack-turn` declares `tokio` with **no `net`** feature and is gated behind the optional `signal` feature (`dep:amplihack-turn`), so default builds don't compile it. Confirmed via `cargo tree`: no `net` in the `amplihack-turn` feature tree. |
| **Sensitive-data handling** | ✅ No secrets logged/committed. UTF-8 decode of stdout uses lossy fallback only on invalid bytes. No credentials, tokens, or env captured. |
| **Concurrency safety** | ✅ `SerialTurnDriver` async-mutex guarantees at most one `copilot --session-id <same>` process in flight, preventing session-state races. |

### Findings

**ℹ️ Info (pre-existing, out of scope) — error-string information disclosure.** `turn.rs:234` embeds the child's combined `stdout`+`stderr` into the non-zero-exit error string, which the chat posts to the group. If the driven `copilot` ever emits sensitive content on failure, it would be surfaced to the group/logs. This is **behaviour-identical to `main`** and correctly untouched by this move-only PR. Flagging only so it can be tracked for a future hardening pass (e.g., redact or cap the embedded output) — **not** a blocker here.

**No P0/P1 security findings.** No injection vectors, no privilege escalation, no new sensitive-data exposure, no new network surface.

**Comment by rysweet** [comment]
## Step 17d — Philosophy Guardian Review

**Verdict: ✅ COMPLIANT — APPROVE.** This PR is a textbook example of amplihack's design philosophy: a ruthlessly simple, self-contained brick with clean studs, extracted behaviour-identically. No blocking (P0) philosophy violations.

### Philosophy compliance checklist

- [x] **Ruthless simplicity achieved.** The new driver core is ~90 lines of actual logic. `run_session_loop` is a single `loop { match next_prompt }` with exactly three arms (`Ready`/`Idle`/`Closed`) — no hidden retry, no timer, no turn cap, no config knobs. The public API is minimal: two traits (`AgentSession`, `Channel`), one enum (`NextPrompt`), one output type, two error enums, one free function. Nothing speculative.
- [x] **Bricks & studs pattern followed.** `amplihack-turn` is a true leaf brick: Signal-independent, self-contained, regeneratable from its doc contract. Its "studs" (the public trait surface + `run_session_loop`) are stable and documented. `amplihack-signal` connects via those studs and nothing else. The dependency is `signal`-feature-gated, so the seam is explicit and optional.
- [x] **Zero-BS implementation.** Scanned all new/moved source: no `todo!`, `unimplemented!`, `panic!`, `FIXME`, faked APIs, or swallowed exceptions. Every failure path surfaces as an explicit `TurnError`/`ChannelError` and propagates via `?` — the module doc's "No silent fallbacks" claim is actually enforced. The default `publish_output` no-op is a **legitimate default** (replay-less channels), not a stub — it returns `Ok(())` by contract, documented as such. The `let _ = read_to_end` / `expect("preempt mutex not poisoned")` sites are **pre-existing** and part of the behaviour-identical move (verified: `turn.rs` differs from `main` only by `super::`→`crate::`).
- [x] **No over-engineering.** Feature footprint is disciplined (crusty condition #4): `amplihack-turn` pulls only `process`, `rt`, `macros`, `time`, `sync`, `io-util` from tokio — **no `net` stack** (verified via `cargo tree`). The TCP transport correctly stays gated inside `amplihack-signal`. No premature generalization: the traits exist because there are genuinely ≥2 agent/channel implementations to unify, not "just in case".
- [x] **Clean module boundaries.** `amplihack-signal` reduces `chat/turn.rs` and `chat/allowlist.rs` to thin re-export shims preserving original paths, so callers and PR-1 characterization tests compile unchanged. The boundary is one-directional (signal → turn), no circular coupling.

### Advisory (non-blocking, consistency with Step 17b code review)

- **P2 (latent):** `run_session_loop` is **new** code (not moved), and its `Idle` arm uses `tokio::task::yield_now().await`. The doc comment asserts the wait "yields cooperatively to the runtime rather than busy-spinning the CPU" — but `yield_now()` does not block; a channel that returns `Idle` repeatedly will hot-poll at ~100% CPU. This is a **correctness/doc-accuracy** gap, not a philosophy violation, and is harmless until a live `Idle`-returning channel is wired. Recommend aligning the implementation with the doc claim (e.g. await a liveness/notify signal) before that happens. Already tracked in the inline code-review comment.

### Evidence
- True PR delta verified against merge-base (branch is behind `main`; GitHub's shown `merge-flow.md` deletion is a stale-branch artifact, not a real change).
- Behaviour-identical move confirmed: `turn.rs` = import-only diff, `allowlist.rs` byte-identical.
- Gates green: `amplihack-turn` 14/14 tests, fmt, clippy, no tokio `net`.

**No philosophy remediation required. Compliant.**

**Comment by rysweet** [comment]
## Step 18b — Review feedback implemented

All three Step 16/17 reviews approved with zero blocking issues. Actioned the triaged plan:

### Addressed in this PR
- **#1 (P2 — Idle busy-spin) — FIXED** (`crates/amplihack-turn/src/lib.rs`)
  - Replaced the `run_session_loop` `Idle` arm's bare `tokio::task::yield_now()` (which only re-queues and can hot-poll at ~100% CPU) with a short bounded `tokio::time::sleep(IDLE_BACKOFF = 5ms)`. The "not busy-spinning" contract now actually holds; still **no** wall-clock timeout on the wait. Doc comment updated to describe the backoff accurately.
  - Gates: `cargo test -p amplihack-turn` **14/14 pass** (incl. `idle_waits_then_processes_pending_prompt`), `fmt --check` pass, `clippy --all-targets` clean, `cargo tree` confirms **no tokio `net`** feature leaks in.
- **#4 (P3 — branch hygiene) — DONE**
  - Rebased onto `main`. This clears the phantom `docs/reference/merge-flow.md` deletion (it was a stale-branch artifact — the file was added on `main` by #1087 after this branch's base). The diff is now clean and move-only + the fix.
  - PR title updated from `Update Cargo.toml with 14 changed files` to a descriptive one.

### Deferred (kept out of this move-only PR)
- **#2 (P3 — version bump 0.12.0→0.13.0):** left as-is; adding a new public crate reasonably justifies a minor bump. No action.
- **#3 (P3 — info disclosure at `turn.rs:234`):** pre-existing, byte-identical to `main`. Filed follow-up **#1092** rather than expanding this clean extraction PR's scope.

Net: one small correctness fix + a rebase. Ready to merge.
